usr/src/uts/common/fs/zfs/dbuf.c

```diff
@@ -1217,6 +1217,11 @@ dbuf_unoverride(dbuf_dirty_record_t *dr)
 	uint64_t txg = dr->dr_txg;
 
 	ASSERT(MUTEX_HELD(&db->db_mtx));
+	/*
+	 * This assert is valid because dmu_sync() expects to be called by
+	 * a zilog's get_data while holding a range lock.  This call only
+	 * comes from dbuf_dirty() callers who must also hold a range lock.
```

Comment:
As far as I can tell, this is true.  It's unfortunate that the DMU is depending on the ZPL's range locks :-(  Thanks for documenting it.